### PR TITLE
Added redis observability to cache

### LINKED ISSUE #266: Open Census for Redis
### TL;DR

Implement https://opencensus.io/integrations/redis/go/

### PR DESCRIPTION
**Pull request by taddari**
Fixes #266

## Proposed Changes

* Instrument functions in pkg/cache/redis with redis opencensus client

**Release Note**

```release-note
Add redis opencensus metrics to the verification server cache.
```

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -38,6 +38,8 @@ import (
 	"github.com/google/exposure-notifications-verification-server/pkg/ratelimit"
 	"github.com/google/exposure-notifications-verification-server/pkg/ratelimit/limitware"
 	"github.com/google/exposure-notifications-verification-server/pkg/render"
+	"github.com/opencensus-integrations/redigo/redis"
+	"go.opencensus.io/stats/view"
```
Comment (icco) [inline cmd/server/main.go:42]: Move these to the last import block

File: cmd/apiserver/main.go [hunk 2]
```diff
@@ -34,11 +38,8 @@ import (
 	"github.com/google/exposure-notifications-verification-server/pkg/ratelimit"
 	"github.com/google/exposure-notifications-verification-server/pkg/ratelimit/limitware"
 	"github.com/google/exposure-notifications-verification-server/pkg/render"
-
-	"github.com/google/exposure-notifications-server/pkg/keys"
-	"github.com/google/exposure-notifications-server/pkg/logging"
-	"github.com/google/exposure-notifications-server/pkg/observability"
-	"github.com/google/exposure-notifications-server/pkg/server"
+	"github.com/opencensus-integrations/redigo/redis"
+	"go.opencensus.io/stats/view"
```
Comment (icco) [inline cmd/apiserver/main.go:42]: Move these to the last import block

File: cmd/apiserver/main.go [hunk 3]
```diff
@@ -24,6 +24,10 @@ import (
 	"os"
 	"strconv"
 
+	"github.com/google/exposure-notifications-server/pkg/keys"
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"github.com/google/exposure-notifications-server/pkg/observability"
+	"github.com/google/exposure-notifications-server/pkg/server"
```
Comment (icco) [inline cmd/apiserver/main.go:30]: add these to the second package block... altho I don't think you need them?

File: cmd/server/main.go (hunk 1 above)
Comment (taddari) [reply cmd/server/main.go]: Done, refactored to proper import block and then sorted the imports

File: cmd/apiserver/main.go (hunk 2 above)
Comment (taddari) [reply cmd/apiserver/main.go]: Done, refactored to proper import block and then sorted the imports

File: cmd/apiserver/main.go (hunk 3 above)
Comment (taddari) [reply cmd/apiserver/main.go:30]: I think those are needed. I just sorted the imports already there. Refactored to proper import block and then sorted the imports

**Comment by sethvargo** [comment]
@taddari I released a new go-limiter and go-redisstore that supports contexts if you wanna pull those in here.

File: pkg/cache/redis.go [hunk 4]
```diff
@@ -273,24 +272,26 @@ func (c *redis) Close() error {
...
 // withConn runs the function with a conn, ensuring cleanup of the connection.
-func (c *redis) withConn(f func(conn redigo.Conn) error) error {
+func (c *redis) withConn(f func(conn redigo.ConnWithContext) error) error {
 	if f == nil {
 		return fmt.Errorf("missing function")
 	}
 
-	conn := c.pool.Get()
+	ctx := context.Background()
+	conn := c.pool.GetWithContext(ctx).(redigo.ConnWithContext)
```
Comment (sethvargo) [inline pkg/cache/redis.go:281]: Do a type check here so we don't panic:

```go
conn, ok := c.pool.GetWithContext(ctx).(redigo.ConnWithContext)
if !ok {
  return fmt.Errorf("redis conn is not ConnWithContext")
}
```

File: pkg/ratelimit/limitware/middleware.go [hunk 5]
```diff
@@ -111,10 +110,6 @@ func NewMiddleware(ctx context.Context, s limiter.Store, f httplimit.KeyFunc, op
 		return nil, fmt.Errorf("stat view registration failure: %w", err)
 	}
 
-	if err := view.Register(redis.ObservabilityMetricViews...); err != nil {
```
Comment (sethvargo) [inline pkg/ratelimit/limitware/middleware.go:114]: Can you help me understand why we moved this out of the main package and put it throughout the server packages? I'd rather leave this in the middleware or observability packages.

File: cmd/adminapi/main.go [hunk 6]
```diff
@@ -74,6 +75,10 @@ func realMain(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("unable to create ObservabilityExporter provider: %w", err)
 	}
+
```
Comment (sethvargo) [inline cmd/adminapi/main.go:78]: I'm not a fan of putting this everywhere - could we do this in the main observability package?

File: pkg/cache/redis.go (hunk 4 above)
Comment (taddari) [reply pkg/cache/redis.go]: Done

File: cmd/adminapi/main.go (hunk 6 above)
Comment (taddari) [reply cmd/adminapi/main.go]: Done, moved inside cache service

File: pkg/ratelimit/limitware/middleware.go (hunk 5 above)
Comment (taddari) [reply pkg/ratelimit/limitware/middleware.go]: You are correct, moving inside server is unnecessary as the servers might not be using redis at all. Moved the code inside the code that uses redis.

**Pull request merged**